1 - import

2 - Data Cleaning

3 - Data Manipulation

4 - export

# Import

In [52]:
import pandas as pd
import numpy as np
import os 

path = r'/Users/amritambe/Desktop/CareerF Python/Insta Cart Analysis'
#orders.csv import and drop column
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'original Data', 'orders.csv'))
df_ords = df_ords.drop(columns = ['eval_set'])
#products.csv import 
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'original Data', 'products.csv'))
#departments.csv import
df_depts = pd.read_csv(os.path.join(path, '02 Data', 'original Data', 'departments.csv'))


In [53]:
#Exploring orders.csv

In [54]:
df_ords.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [55]:
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

# Data Cleaning

In [56]:
#renaming orders column

In [57]:
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [58]:
df_ords.dtypes

order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [59]:
#changing the data type of ID to string

In [60]:
df_ords[['order_id', 'user_id']] = df_ords[['order_id', 'user_id']].astype(str)

In [61]:
df_ords.dtypes

order_id                   object
user_id                    object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [62]:
#transposing and manipulating header of departments.csv

In [63]:
df_depts_T = df_depts.T
df_depts_T.head(10)

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta

In [64]:
df_depts_T.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [65]:
new_header = df_depts_T.iloc[0]
new_header

0    department
Name: department_id, dtype: object

In [66]:
df_depts_T_new = df_depts_T[1:]
df_depts_T_new.head()

0
1   frozen
2    other
3   bakery
4  produce
5  alcohol

In [67]:
df_depts_T_new.columns = new_header

In [68]:
df_depts_T_new.tail()

department_id department
17             household
18                babies
19                snacks
20                  deli
21               missing

In [69]:
#creating a dictionary from departments

In [70]:
new_dict = df_depts_T_new.to_dict('index')

In [71]:
new_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [72]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [74]:
print(new_dict.get('19'))

{'department': 'snacks'}


In [75]:
#make a new dataframe with only dept 19, snacks

In [77]:
df_snacks = df_prods[df_prods['department_id'] == 19]
df_snacks.head(15)

product_id                                 product_name  aisle_id  \
0             1                   Chocolate Sandwich Cookies        61   
15           16                Mint Chocolate Flavored Syrup       103   
24           25      Salted Caramel Lean Protein & Fiber Bar         3   
31           32                Nacho Cheese White Bean Chips       107   
40           41  Organic Sourdough Einkorn Crackers Rosemary        78   
55           56                   Healthy Pop Butter Popcorn        23   
58           59         Medium Taqueria Style Chipotle Salsa        50   
76           77            Coconut Chocolate Chip Energy Bar         3   
77           78            Nutter Butter Cookie Bites Go-Pak        61   
92           93                  Uncured Cracked Pepper Beef        23   
101         102                        Danish Butter Cookies        61   
102         103                        Sprouted Kale Cracker        78   
114         115            Scooby-Doo! Fruit Flavored Snacks        50   
123         124                      Sun Cups Dark Chocolate        45   
134         135             Dark Chocolate Ice Cream Topping       103   

     department_id  prices  
0               19     5.8  
15              19     5.2  
24              19     1.9  
31              19     4.9  
40              19     6.5  
55              19     4.4  
58              19     6.8  
76              19     6.4  
77              19     3.0  
92              19     2.4  
101             19     5.5  
102             19     6.4  
114             19     2.7  
123             19     3.0  
134             19     2.4

In [ ]:
#Answers to questions: 4.4

In [ ]:
# Find another identifier variable in the df_ords dataframe that doesn’t need to be included 
# in your analysis as a numeric variable and change it to a suitable format.

In [78]:
#this is done above in this line:
# df_ords[['order_id', 'user_id']] = df_ords[['order_id', 'user_id']].astype(str)

In [ ]:
# Look for a variable in your df_ords dataframe with an unintuitive name 
# and change its name without overwriting the dataframe.

In [80]:
df_ords.rename(columns = {'days_since_prior_order' : 'days_between_last_order'}, inplace = True)
df_ords.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_between_last_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0

In [ ]:
# Your client wants to know what the busiest hour is for placing orders. 
# Find the frequency of the corresponding variable and share your findings.

In [82]:
df_ords['order_hour_of_day'].value_counts()

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

The most common hour of the day is 10am with 288,418 transactions.

In [ ]:
# Determine the meaning behind a value of 4 in the "department_id" 
# column within the df_prods dataframe using a data dictionary.


In [92]:
new_dict.get('4')

{'department': 'produce'}

The department with id = 4 is produce.

# Data Manipulation

In [ ]:
# The sales team in your client’s organization wants to 
# know more about breakfast item sales. Create a subset containing only the required information.


In [93]:
df_breakfast = df_prods[df_prods['department_id']==14]
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [95]:
# They’d also like to see details about products that customers
# might use to throw dinner parties. Your task is to find all observations 
# from the entire dataframe that include items from the following departments: 
# alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.


In [104]:
new_dict
#the department ids i need are 5, 7, 12, and 20
#df will be 'df_dinner_parties'

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [103]:
df_dinner_parties = df_prods[df_prods['department_id'].isin([5, 7, 12, 20])]
df_dinner_parties

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

In [ ]:
# It’s important that you keep track of total counts in your dataframes. 
# How many rows does the last dataframe you created have?


The df_dinner_parties dataframe has 7650 rows. 

In [ ]:
# Someone from the data engineers team in Instacart thinks they’ve spotted something 
# strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [106]:
df_cust_one = df_ords[df_ords['user_id']=='1']
df_cust_one

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_between_last_order  
0                       NaN  
1                      15.0  
2                      21.0  
3                      29.0  
4                      28.0  
5                      19.0  
6                      20.0  
7                      14.0  
8                       0.0  
9                      30.0  
10                     14.0

In [ ]:
# You also need to provide some details about this user’s behavior. 
# What basic stats can you provide based on the information you have?

In [107]:
df_cust_one.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_between_last_order  
count                10.000000  
mean                 19.000000  
std                   9.030811  
min                   0.000000  
25%                  14.250000  
50%                  19.500000  
75%                  26.250000  
max                  30.000000

Customer has placed 11 orders, 3 on sunday, 2 on monday, 2 on tuesday, and 4 on wednesday. 
Average days between orders is 19. The Average hour the orders are placed is 10am.


In [ ]:
# Check the organization and structure of your notebook. 
# Be sure to include section headings and code comments.

# export

In [ ]:
# Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [108]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

In [ ]:
# Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your 
#“Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [110]:
df_depts_T_new.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'departments_wrangled.csv'))